src/grader/test_lib.ml

```diff
@@ -358,421 +400,439 @@ module Make
-          let before = !variables in
-          let modules_before = !modules in
-          (match vexpr with
-           | Some vexpr -> ignore (mapper.expr mapper vexpr)
-           | None -> ()) ;
-          (match label with
-           | Asttypes.Nolabel -> ()
-           | Asttypes.Labelled label
-           | Asttypes.Optional label ->
+           match List.rev txt with
+           | [] -> ()
+           | _ :: prefixes -> all prefixes in
+      let treat_module ({ Location.txt; _ } as ident) =
+        treat_module_prefixes ident ;
+        match Longident.flatten txt with
+        | [ m ] | m :: _  when List.mem m !modules (* shadowed *) -> ()
+        | _ -> add @@ on_module_occurence (name ident) in
+      let treat_variable ({ Location.txt; _ } as ident) =
+        treat_module_prefixes ident ;
+        match Longident.flatten txt with
+        | m :: _ :: _  when List.mem m !modules (* shadowed *) -> ()
+        | [ v ] when List.mem v !variables (* shadowed *) -> ()
+        | _ -> add @@ on_variable_occurence (name ident) in
+      let expr mapper expr =
+        add @@ on_expression expr ;
+        match expr with
+        | { pexp_desc = Pexp_open (popen_override, popen_lid, iexpr); _ } ->
+           let o = { popen_lid ; popen_override ;
+                     popen_loc = Location.none ;
+                     popen_attributes = [] } in
+           let before = !modules in
+           add @@ on_open o ;
+           treat_module popen_lid ;
+           variables := [] ;
+           modules := [] (* over approximation *) ;
+           ignore (mapper.expr mapper iexpr) ;
+           modules := before ;
+           expr
+        | { pexp_desc = Pexp_letmodule ({ Location.txt = name; _ }, mexpr, iexpr); _ } ->
+           let before = !modules in
+           let variables_before = !variables in
+           ignore (mapper.module_expr mapper mexpr) ;
+           modules := name :: before ;
+           variables := variables_before ;
+           ignore (mapper.expr mapper iexpr) ;
+           modules := before ;
+           expr
+        | { pexp_desc = Pexp_apply (fn, args); _ } as e ->
+           let args = List.map
+                        (function
+                         | (Asttypes.Nolabel, v) -> ("", v)
+                         | ((Asttypes.Labelled n | Asttypes.Optional n), v) -> (n, v))
+                        args in
+           add @@ on_function_call (fn, args) ;
+           default_mapper.expr mapper e
+        | { pexp_desc = Pexp_ident ident; _ } as e ->
+           treat_variable ident ;
+           default_mapper.expr mapper e
+        | { pexp_desc = Pexp_construct (ident, _); _ } as e ->
+           treat_module_prefixes ident ;
+           default_mapper.expr mapper e
+        | { pexp_desc = Pexp_record (fields, _); _ } as e ->
+           List.iter (fun (ident, _) -> treat_module_prefixes ident) fields ;
+           default_mapper.expr mapper e
+        | { pexp_desc = Pexp_field (_, ident); _ } as e ->
```

Comment:
The next three cases could be factorized out with a disjunctive pattern, right?

Reply:
Well, it is not really relative to documentation. Should I do that in another branch ?

Reply:
You are right: forget about this comment.